### PR TITLE
ARROW-16850: [C++] Copy CSV data field and end chars separately

### PR DESCRIPTION
**Pull request by cyb70289**
It's a finding when evaluating Arm Statistical Profiling Extension(SPE)
on Neoverse-N1 with Arrow CSV writer benchmark. With SPE, we are able to
locate precisely the machine code causing heavy cache miss or the exact
branch suffering from high misprediction rate.

CSV writer benchmark reveals relatively high branch misprediction rate
inside glibc `memcpy` function. The branch is comparing the buffer size
against 8, and to run different copying code per comparison result.

Arrow CSV writer populates one column at once. In the inner loop, it
does two `memcpy`. First one is to copy the field value(string) from
Arrow string array. Then it does another copy to append the delimiter
or end-of-line(if last column), which is at most 2 chars. Data in same
column is normally of similar size, but as we are copying a very short
delimiter after each data field, looks it makes CPU harder to predict
the correct code path in `memcpy` as input data size varies quickly.

This PR adds a trivial `CopyEndChars` routine to copy only the
delimiter and end-of-line. It improves performance significantly on
Arm Neoverse-N1. Skylake also sees mild improvement.

### CODE REVIEW

**Comment by cyb70289** [comment]
**neoverse-n1, gcc-10**
```
$ archery benchmark diff --suite-filter="arrow-csv-writer-benchmark" --cc=gcc-10 --cxx=g++-10
-----------------------------------------------------------------------
Non-regressions: (20)
-----------------------------------------------------------------------
                   benchmark         baseline       contender  change %
     WriteCsvStringNoQuote/0    1.257 GiB/sec   1.875 GiB/sec    49.220
 WriteCsvStringRejectQuote/0    1.126 GiB/sec   1.595 GiB/sec    41.646
     WriteCsvStringNoQuote/1    1.186 GiB/sec   1.680 GiB/sec    41.622
    WriteCsvStringNoQuote/10 1006.623 MiB/sec   1.321 GiB/sec    34.365
 WriteCsvStringRejectQuote/1    1.066 GiB/sec   1.430 GiB/sec    34.056
   WriteCsvStringWithQuote/1  377.222 MiB/sec 492.581 MiB/sec    30.581
    WriteCsvStringNoQuote/50  475.139 MiB/sec 584.854 MiB/sec    23.091
WriteCsvStringRejectQuote/10  927.789 MiB/sec   1.107 GiB/sec    22.193
WriteCsvStringRejectQuote/50  450.702 MiB/sec 536.354 MiB/sec    19.004
 WriteCsvNumericCheckQuote/0  219.353 MiB/sec 248.296 MiB/sec    13.195
 WriteCsvNumericCheckQuote/1  212.097 MiB/sec 236.551 MiB/sec    11.529
           WriteCsvNumeric/0  164.644 MiB/sec 182.507 MiB/sec    10.850
WriteCsvNumericCheckQuote/50  114.087 MiB/sec 125.188 MiB/sec     9.731
WriteCsvNumericCheckQuote/10  191.798 MiB/sec 210.442 MiB/sec     9.721
  WriteCsvStringWithQuote/50  304.072 MiB/sec 333.337 MiB/sec     9.624
           WriteCsvNumeric/1  159.534 MiB/sec 173.415 MiB/sec     8.700
          WriteCsvNumeric/50  100.055 MiB/sec 108.291 MiB/sec     8.232
   WriteCsvStringWithQuote/0  474.221 MiB/sec 507.793 MiB/sec     7.080
  WriteCsvStringWithQuote/10  434.169 MiB/sec 460.740 MiB/sec     6.120
          WriteCsvNumeric/10  149.530 MiB/sec 156.290 MiB/sec     4.521
```

**neoverse-n1, clang-12**
```
$ archery benchmark diff --suite-filter="arrow-csv-writer-benchmark" --cc=clang-12 --cxx=clang++-12
----------------------------------------------------------------------
Non-regressions: (20)
----------------------------------------------------------------------
                   benchmark        baseline       contender  change %
 WriteCsvStringRejectQuote/0   1.089 GiB/sec   1.700 GiB/sec    56.186
 WriteCsvStringRejectQuote/1   1.016 GiB/sec   1.524 GiB/sec    49.957
     WriteCsvStringNoQuote/0   1.261 GiB/sec   1.825 GiB/sec    44.696
WriteCsvStringRejectQuote/10 876.558 MiB/sec   1.213 GiB/sec    41.658
     WriteCsvStringNoQuote/1   1.165 GiB/sec   1.573 GiB/sec    35.065
WriteCsvStringRejectQuote/50 425.765 MiB/sec 565.216 MiB/sec    32.753
    WriteCsvStringNoQuote/10 999.100 MiB/sec   1.290 GiB/sec    32.178
    WriteCsvStringNoQuote/50 460.679 MiB/sec 592.269 MiB/sec    28.564
 WriteCsvNumericCheckQuote/0 227.334 MiB/sec 267.121 MiB/sec    17.501
 WriteCsvNumericCheckQuote/1 217.809 MiB/sec 252.958 MiB/sec    16.138
           WriteCsvNumeric/0 171.684 MiB/sec 198.768 MiB/sec    15.775
WriteCsvNumericCheckQuote/10 195.498 MiB/sec 223.550 MiB/sec    14.349
WriteCsvNumericCheckQuote/50 111.831 MiB/sec 127.825 MiB/sec    14.302
          WriteCsvNumeric/10 149.899 MiB/sec 171.253 MiB/sec    14.245
           WriteCsvNumeric/1 166.593 MiB/sec 189.017 MiB/sec    13.460
          WriteCsvNumeric/50  96.724 MiB/sec 108.630 MiB/sec    12.310
  WriteCsvStringWithQuote/50 282.800 MiB/sec 315.637 MiB/sec    11.612
   WriteCsvStringWithQuote/1 425.883 MiB/sec 449.183 MiB/sec     5.471
  WriteCsvStringWithQuote/10 401.068 MiB/sec 421.065 MiB/sec     4.986
   WriteCsvStringWithQuote/0 441.107 MiB/sec 422.944 MiB/sec    -4.118
```

**Comment by cyb70289** [comment]
**skylake, gcc-10**
```
numactl -m 1 -N 1 archery benchmark diff --suite-filter="arrow-csv-writer-benchmark" --cc=gcc-10 --cxx=g++-10
-----------------------------------------------------------------------
Non-regressions: (20)
-----------------------------------------------------------------------
                   benchmark        baseline       contender  change %
 WriteCsvStringRejectQuote/0   1.483 GiB/sec   1.883 GiB/sec    26.948
 WriteCsvStringRejectQuote/1   1.283 GiB/sec   1.558 GiB/sec    21.469
     WriteCsvStringNoQuote/0   1.539 GiB/sec   1.855 GiB/sec    20.591
  WriteCsvStringWithQuote/10 316.582 MiB/sec 366.179 MiB/sec    15.666
    WriteCsvStringNoQuote/10   1.052 GiB/sec   1.182 GiB/sec    12.342
    WriteCsvStringNoQuote/50 463.062 MiB/sec 516.478 MiB/sec    11.535
     WriteCsvStringNoQuote/1   1.411 GiB/sec   1.552 GiB/sec    10.028
WriteCsvStringRejectQuote/10   1.007 GiB/sec   1.081 GiB/sec     7.396
           WriteCsvNumeric/0 169.573 MiB/sec 180.847 MiB/sec     6.649
  WriteCsvStringWithQuote/50 276.665 MiB/sec 292.947 MiB/sec     5.885
 WriteCsvNumericCheckQuote/0 227.048 MiB/sec 239.672 MiB/sec     5.560
 WriteCsvNumericCheckQuote/1 213.929 MiB/sec 225.795 MiB/sec     5.546
WriteCsvNumericCheckQuote/10 185.902 MiB/sec 196.156 MiB/sec     5.516
WriteCsvNumericCheckQuote/50 102.471 MiB/sec 107.432 MiB/sec     4.842
           WriteCsvNumeric/1 160.339 MiB/sec 167.057 MiB/sec     4.190
          WriteCsvNumeric/10 142.298 MiB/sec 145.386 MiB/sec     2.170
   WriteCsvStringWithQuote/1 452.774 MiB/sec 457.586 MiB/sec     1.063
WriteCsvStringRejectQuote/50 451.117 MiB/sec 454.032 MiB/sec     0.646
   WriteCsvStringWithQuote/0 482.822 MiB/sec 485.438 MiB/sec     0.542
          WriteCsvNumeric/50  89.502 MiB/sec  89.804 MiB/sec     0.338
```

**skylake, clang-12**
```
$ numactl -m 1 -N 1 archery benchmark diff --suite-filter="arrow-csv-writer-benchmark" --cc=clang-12 --cxx=clang++-12
-----------------------------------------------------------------------
Non-regressions: (20)
-----------------------------------------------------------------------
                   benchmark         baseline       contender  change %
    WriteCsvStringNoQuote/10  875.541 MiB/sec   1.002 GiB/sec    17.226
 WriteCsvStringRejectQuote/1    1.275 GiB/sec   1.479 GiB/sec    16.008
 WriteCsvStringRejectQuote/0    1.414 GiB/sec   1.600 GiB/sec    13.141
     WriteCsvStringNoQuote/1    1.296 GiB/sec   1.464 GiB/sec    12.996
WriteCsvStringRejectQuote/10 1013.392 MiB/sec   1.109 GiB/sec    12.031
WriteCsvStringRejectQuote/50  440.191 MiB/sec 473.814 MiB/sec     7.638
   WriteCsvStringWithQuote/0  395.423 MiB/sec 422.366 MiB/sec     6.814
     WriteCsvStringNoQuote/0    1.566 GiB/sec   1.669 GiB/sec     6.542
 WriteCsvNumericCheckQuote/0  238.286 MiB/sec 247.916 MiB/sec     4.041
           WriteCsvNumeric/0  176.157 MiB/sec 182.788 MiB/sec     3.764
           WriteCsvNumeric/1  167.941 MiB/sec 173.746 MiB/sec     3.457
 WriteCsvNumericCheckQuote/1  223.360 MiB/sec 230.066 MiB/sec     3.002
          WriteCsvNumeric/10  149.428 MiB/sec 153.581 MiB/sec     2.779
  WriteCsvStringWithQuote/10  387.863 MiB/sec 398.526 MiB/sec     2.749
          WriteCsvNumeric/50   92.557 MiB/sec  94.585 MiB/sec     2.192
  WriteCsvStringWithQuote/50  268.692 MiB/sec 274.562 MiB/sec     2.185
WriteCsvNumericCheckQuote/10  191.414 MiB/sec 192.849 MiB/sec     0.750
   WriteCsvStringWithQuote/1  418.052 MiB/sec 404.351 MiB/sec    -3.277
    WriteCsvStringNoQuote/50  431.535 MiB/sec 416.108 MiB/sec    -3.575
WriteCsvNumericCheckQuote/50  103.572 MiB/sec  99.181 MiB/sec    -4.240
```

**Comment by cyb70289** [comment]
@ursabot please benchmark lang=cpp

**Comment by cyb70289** [comment]
@ursabot please benchmark lang=C++

**Comment by cyb70289** [comment]
From conbench, csv writer benchmark improvement on `i9` is about 8% to 25%. `m1` improvement is small (5% to 10%).

File: cpp/src/arrow/csv/writer.cc [hunk 1]
```diff
@@ -59,6 +59,18 @@ namespace csv {
 
 namespace {
 
+// Copy delimiter or end-of-line chars. Though configurable,
+// delimiter is always one char, end-of-line is often at most two chars.
+void copy_endchars(char* dest, const char* src, size_t size) {
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:64]: Nit: `CopyEndChars`?

File: cpp/src/arrow/csv/writer.cc [hunk 2]
```diff
@@ -59,6 +59,18 @@ namespace csv {
 
 namespace {
 
+// Copy delimiter or end-of-line chars. Though configurable,
+// delimiter is always one char, end-of-line is often at most two chars.
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:63]: Explain that this is for performance, stating the reason doing things this way is faster?

**Review by pitrou — COMMENTED** [review]
LGTM, just two nits

File: cpp/src/arrow/csv/writer.cc [hunk 3]
```diff
@@ -59,6 +59,25 @@ namespace csv {
...
 namespace {
 
+// This function is to improve performance. It copies csv delimiter and eol
+// without calling `memcpy`.
+// Each csv field is followed by a delimiter or eol, which is often only one
+// or two chars. If copying both the field and delimiter with `memcpy`, cpu
+// may suffer from high branch misprediction as we are tripping `memcpy` with
+// interleaved (normal/tiny/normal/tiny/...) buffer sizes, which are handled
+// separately inside `memcpy`. This function goes fast path if the buffer
+// size is one or two chars to leave `memcpy` only for copyig csv fields.
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:69]:
```suggestion
// This function is to improve performance. It copies CSV delimiter and eol
// without calling `memcpy`.
// Each CSV field is followed by a delimiter or eol, which is often only one
// or two chars. If copying both the field and delimiter with `memcpy`, CPU
// may suffer from high branch misprediction as we are tripping `memcpy` with
// interleaved (normal/tiny/normal/tiny/...) buffer sizes, which are handled
// separately inside `memcpy`. This function goes fast path if the buffer
// size is one or two chars to leave `memcpy` only for copying CSV fields.
```

**Pull request merged**